pymtl3/stdlib/rtl/queues.py

```diff
@@ -86,11 +86,298 @@ def construct( s, num_entries=2 ):
+                  s.count
+
+#-------------------------------------------------------------------------
+# NormalQueueRTL
+#-------------------------------------------------------------------------
+
+class NormalQueueRTL( Component ):
+
+  def construct( s, EntryType, num_entries=2 ):
+
+    # Interface
+
+    s.enq   = EnqIfcRTL( EntryType )
+    s.deq   = DeqIfcRTL( EntryType )
+    s.count = OutPort( mk_bits( clog2( num_entries ) ) )
+
+    # Components
+
+    s.ctrl  = NormalQueueCtrlRTL ( num_entries )
+    s.dpath = NormalQueueDpathRTL( EntryType, num_entries )
+
+    # Connect ctrl to data path
+
+    s.connect( s.ctrl.wen,     s.dpath.wen     )
+    s.connect( s.ctrl.waddr,   s.dpath.waddr   )
+    s.connect( s.ctrl.raddr,   s.dpath.raddr   )
+
+    # Connect to interface
+
+    s.connect( s.enq.en,  s.ctrl.enq_en   )
+    s.connect( s.enq.rdy, s.ctrl.enq_rdy  )
+    s.connect( s.deq.en,  s.ctrl.deq_en   )
+    s.connect( s.deq.rdy, s.ctrl.deq_rdy  )
+    s.connect( s.count,   s.ctrl.count    )
+    s.connect( s.enq.msg, s.dpath.enq_msg )
+    s.connect( s.deq.msg, s.dpath.deq_msg )
+
+  # Line trace
+
+  def line_trace( s ):
+    return "{}({}){}".format( s.enq, s.count, s.deq )
+
+#-------------------------------------------------------------------------
+# Ctrl for PipeQueue
+#-------------------------------------------------------------------------
+
+class PipeQueueCtrlRTL( Component ):
+
+  def construct( s, num_entries=2 ):
+
+    # Constants
+
+    s.num_entries = num_entries
+    s.last_idx    = num_entries - 1
+    addr_nbits    = clog2( num_entries   )
+    count_nbits   = clog2( num_entries+1 )
+    PtrType       = mk_bits( addr_nbits  )
+    CountType     = mk_bits( count_nbits )
+
+    # Interface
+
+    s.enq_en  = InPort ( Bits1   )
+    s.enq_rdy = OutPort( Bits1   )
+    s.deq_en  = InPort ( Bits1   )
+    s.deq_rdy = OutPort( Bits1   )
+    s.count   = OutPort( PtrType )
+
+    s.wen     = OutPort( Bits1   )
+    s.waddr   = OutPort( PtrType )
+    s.raddr   = OutPort( PtrType )
+
+    # Registers
+
+    s.head = Wire( PtrType )
+    s.tail = Wire( PtrType )
+
+    # Wires
+
+    s.enq_xfer  = Wire( Bits1   )
+    s.deq_xfer  = Wire( Bits1   )
+    s.head_next = Wire( PtrType )
+    s.tail_next = Wire( PtrType )
+
+    # Connections
+
+    s.connect( s.wen,   s.enq_xfer )
+    s.connect( s.waddr, s.tail     )
+    s.connect( s.raddr, s.head     )
+
+    @s.update
+    def up_rdy_signals():
+      if ~s.reset:
+        s.deq_rdy = b1(1) if s.count > 0 else b1(0)
+      else:
+        s.deq_rdy = b1(0)
+
+    @s.update
+    def up_enq_rdy():
+      if ~s.reset:
+        s.enq_rdy = s.count < s.num_entries or s.deq_en
+      else:
+        s.enq_rdy = b1(0)
+
+
+    @s.update
+    def up_xfer_signals():
+      s.enq_xfer  = s.enq_en and s.enq_rdy
+      s.deq_xfer  = s.deq_en and s.deq_rdy
+
+    @s.update
+    def up_next():
+      s.head_next = s.head - b1(1) if s.head > 0 else PtrType( s.last_idx )
```

Comment:
do we need multi-entry pipe and bypass queues? let me bring this up on slack.

Reply:
we have these in cl, so I think we should have their RTL equivalence in our stdlib as well.

Reply:
I was just thinking about the discussion between @cbatten and I on whether a pipe queue with >1 entry is useful. The conclusion was that >=2 entry normal queue is good enough for most cases to buffer the message without elongating the ready path, and we would usually just go for a big normal queue to cut both ready and valid. >=2 entry bypass queue's behavior is a little bit weird -- it can only bypass when the queue is empty. Otherwise the queue acts like a normal queue. We do use a 2-entry bypass queue in ece4750 processor to sustain 1-fetch per cycle during squashing.